### PR TITLE
Change roles

### PR DESCRIPTION
**Pull request by fortmarek**
### Short description 📝

This PR adds the functionality of actually changing users' roles on the backend.

### Checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase.
- [ ] The changes have been tested following the [documented guidelines](https://docs.tuist.io/contributors/testing-strategy/).
- [ ] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [ ] In case the PR introduces changes that affect users, the documentation has been updated.
- [ ] In case the PR introduces changes that affect how the cache artifact is generated starting from the `TuistGraph.Target`, the `Constants.cacheVersion` has been updated.

### CODE REVIEW

**Comment by fortmarek** [comment]
Updated the code to use `pundit` instead of `cancan`.

File: projects/cloud/app/graphql/mutations/change_user_role.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+module Mutations
+  class ChangeUserRole < ::Mutations::BaseMutation
+    argument :user_id, ID, required: true
+    argument :organization_id, String, required: true
+    argument :current_role, Types::RoleType, required: true
```
Comment (pepicrft) [inline projects/cloud/app/graphql/mutations/change_user_role.rb:5]: Because the `current_role` data is already in `user` (fetched from `user_id`), I don't think we need to pass this from the client.

File: projects/cloud/app/graphql/mutations/change_user_role.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    argument :new_role, Types::RoleType, required: true
```
Comment (pepicrft) [inline projects/cloud/app/graphql/mutations/change_user_role.rb:6]: If we remove `current_role`, we can rename this one to `role` because it's obvious from the name of the mutation that the role is the one you are trying to set.

File: projects/cloud/app/graphql/types/query_type.rb [hunk 3]
```diff
@@ -39,7 +39,7 @@ def accounts
       argument :account_name, String, required: true
     end
     def project(name:, account_name:)
-      ProjectFetchService.call(name: name, account_name: account_name)
+      ProjectFetchService.call(name: name, account_name: account_name, current_user: context[:current_user])
```
Comment (pepicrft) [inline projects/cloud/app/graphql/types/query_type.rb:42]: Nitpick: I think the name of the parameter should be `user:` instead of `current:`. While `current` makes sense in the context of a controller to represent the authenticated user, I think the service doesn't have that context. Give me the user and I'll give you the projects the user has access to.

File: projects/cloud/app/graphql/types/role_type.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+module Types
+  class RoleType < Types::BaseEnum
+    value "admin"
```
Comment (pepicrft) [inline projects/cloud/app/graphql/types/role_type.rb:5]: Correct me but, this then means we have 3 roles: admin, user, and owner of the account?

File: projects/cloud/app/services/change_user_role_service.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    @current_user = current_user
+  end
+
+  def call
+    return user if current_role == new_role
+    ActiveRecord::Base.transaction do
+      organization = Organization.find(organization_id)
+      raise Error::Unauthorized unless OrganizationPolicy.new(current_user, organization).update?
+      user = User.find(user_id)
+      user.remove_role(current_role, organization)
```
Comment (pepicrft) [inline projects/cloud/app/services/change_user_role_service.rb:25]: I think we should design this service to take whatever role the user has to the new role. We can fetch the roles from the user that we fetched from the database.

File: projects/cloud/app/services/change_user_role_service.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    @current_role = current_role
+    @new_role = new_role
+    @current_user = current_user
+  end
+
+  def call
+    return user if current_role == new_role
+    ActiveRecord::Base.transaction do
+      organization = Organization.find(organization_id)
+      raise Error::Unauthorized unless OrganizationPolicy.new(current_user, organization).update?
```
Comment (pepicrft) [inline projects/cloud/app/services/change_user_role_service.rb:23]: We need to catch this one at the mutation level and convert that into a standard GraphQL response for errors.

File: projects/cloud/app/policies/organization_policy.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+class OrganizationPolicy < ApplicationPolicy
```
Comment (pepicrft) [inline projects/cloud/app/policies/organization_policy.rb:1]: Beautiful. It's so easy to read.

File: projects/cloud/app/policies/project_policy.rb [hunk 8]
```diff
@@ -0,0 +1,9 @@
+class ProjectPolicy < ApplicationPolicy
+  def show?
```
Comment (pepicrft) [inline projects/cloud/app/policies/project_policy.rb:2]: Same here, it's very easy to read.

File: projects/cloud/app/services/project_fetch_service.rb [hunk 9]
```diff
@@ -1,16 +1,23 @@
...
     @name = name
     @account_name = account_name
+    @current_user = current_user
   end
 
   def call
     account = Account.find_by(name: account_name)
-    Project.find_by(account_id: account.id, name: name)
+    project = Project.find_by(account_id: account.id, name: name)
+    raise Error::Unauthorized unless ProjectPolicy.new(current_user, project).show?
```
Comment (pepicrft) [inline projects/cloud/app/services/project_fetch_service.rb:20]: Make sure you rescue from this one at the mutation level and return a standard response agreed with the frontend.

**Review by pepicrft — APPROVED** [review]

File: projects/cloud/app/graphql/types/role_type.rb (hunk 4 above)
Comment (fortmarek) [reply projects/cloud/app/graphql/types/role_type.rb:5]: Yes, I'm for adding `owner` in a different PR, though. There are some intricacies tied to changing it and right now, `admin` == `owner` in terms of what their capabilities.

File: projects/cloud/app/services/change_user_role_service.rb (hunk 5 above)
Comment (fortmarek) [reply projects/cloud/app/services/change_user_role_service.rb:25]: Could you rephrase what you mean?

Currently, roles for a given `organization` are exclusionary - a `User` model can only have a role of an `admin`. `User` cannot be `admin` _and_ `user` at the same time. The other approach could be `User` would have `user` _and_ `admin` role. I am not 100 % sure what's the preferred practice, I've seen both approaches when searching for this.

File: projects/cloud/app/services/project_fetch_service.rb (hunk 9 above)
Comment (fortmarek) [reply projects/cloud/app/services/project_fetch_service.rb]: good point, done!

File: projects/cloud/app/graphql/mutations/change_user_role.rb [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+    argument :organization_id, String, required: true
+    argument :role, Types::RoleType, required: true
+
+    type Types::UserType
+
+    def resolve(attributes)
+      begin
+        ChangeUserRoleService.call(**attributes, acting_user: context[:current_user])
+      rescue ChangeUserRoleService::Error::Unauthorized
+        raise GraphQL::ExecutionError, "You do not have a permission to change a role for this user."
```
Comment (pepicrft) [inline projects/cloud/app/graphql/mutations/change_user_role.rb:13]: I'm afraid this will result in a non-standard error format that might change if `ruby-graphq` changes the way errors are managed. I'd instead introduce a new type that models errors, and return that instead. I'd leave `GraphQL::ExecutionError` only for uncaught errors.

File: projects/cloud/app/graphql/mutations/change_user_role.rb [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+  class ChangeUserRole < ::Mutations::BaseMutation
+    argument :user_id, ID, required: true
+    argument :organization_id, String, required: true
+    argument :role, Types::RoleType, required: true
+
+    type Types::UserType
+
+    def resolve(attributes)
+      begin
+        ChangeUserRoleService.call(**attributes, acting_user: context[:current_user])
```
Comment (pepicrft) [inline projects/cloud/app/graphql/mutations/change_user_role.rb:11]: Is `acting_user` the user trying to change the role? If so, what about calling it `role_changer`?

File: projects/cloud/app/graphql/types/query_type.rb [hunk 12]
```diff
@@ -39,7 +39,11 @@ def accounts
       argument :account_name, String, required: true
     end
     def project(name:, account_name:)
-      ProjectFetchService.call(name: name, account_name: account_name)
+      begin
+        ProjectFetchService.call(name: name, account_name: account_name, user: context[:current_user])
+      rescue ProjectFetchService::Error::Unauthorized
+        raise GraphQL::ExecutionError, "You do not have a permission to view a project #{account_name}/#{name}"
```
Comment (pepicrft) [inline projects/cloud/app/graphql/types/query_type.rb:45]: Similar comment here in regards to error management.

File: projects/cloud/app/services/change_user_role_service.rb [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+  def initialize(user_id:, organization_id:, role:, acting_user:)
+    super()
+    @user_id = user_id
+    @organization_id = organization_id
+    @role = role
+    @acting_user = acting_user
+  end
+
+  def call
+    user = User.find(user_id)
```
Comment (pepicrft) [inline projects/cloud/app/services/change_user_role_service.rb:19]: This will [raise an error](https://api.rubyonrails.org/v6.1.4/classes/ActiveRecord/RecordNotFound.html) if the user cannot be found. I'd rescue from it, and return a custom error that we can then map to an error response in the GraphQL layer. I'd then add a unit test for it.

File: projects/cloud/app/services/project_fetch_service.rb [hunk 14]
```diff
@@ -1,16 +1,23 @@
...
     super()
     @name = name
     @account_name = account_name
+    @user = user
   end
 
   def call
     account = Account.find_by(name: account_name)
-    Project.find_by(account_id: account.id, name: name)
+    project = Project.find_by(account_id: account.id, name: name)
```
Comment (pepicrft) [inline projects/cloud/app/services/project_fetch_service.rb:19]: This might return `nil` and you'll be passing that down to `ProjectPolicy`. I'd instead use `find`, rescue from [this error](https://api.rubyonrails.org/v6.1.4/classes/ActiveRecord/RecordNotFound.html), and return a `Error::ProjectNotFound` error.

**Review by pepicrft — APPROVED** [review]
Some final comments. I think we can merge after that. I think it's important that we make the error responses from the API standard such that we can have also standard error management on the client-side. Solid work Marek! Your ability to learn fast never ceases to impress me.

**Comment by fortmarek** [comment]
I have updated the error handling. In the end, it still boils down to `GraphQL::ExecutionError`, however, I believe raising this error is recommended for the `ruby-graphq` framework. What I did not like about the previous implementation was that the messages were defined in a different file from the errors themselves. 

I am not sure if the solution I used is really idiomatic Ruby, so I'm open to suggestions.

**Review by pepicrft — APPROVED** [review]
